clean up GIL usuage

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32748 clean up GIL usuage**

This is to follow up PR #30630, we need to have GIL when calling jit::toPyObject(), for some binded functions need to be taged with GIL release if underneath C++ codes requires GIL. so
1. pyRef::to_here() and pyRef::local_value() added GIL
2. pyRef::pickle and pyRef::unpickle() added GIL release tag
3. in request_callback_impl, also added GIL as needed
4. for typeParser, use cached jitCompilationUnit_, also clean it up in cleanUp() function

Differential Revision: [D19612337](https://our.internmc.facebook.com/intern/diff/D19612337/)

zhaojuanmao committed:
clean up GIL usuage

This is to follow up PR #30630, we need to have GIL when calling jit::toPyObject(), for some binded functions need to be taged with GIL release if underneath C++ codes requires GIL. so
1. pyRef::to_here() and pyRef::local_value() added GIL
2. pyRef::pickle and pyRef::unpickle() added GIL release tag
3. in request_callback_impl, also added GIL as needed
4. for typeParser, use cached jitCompilationUnit_, also clean it up in cleanUp() function

Differential Revision: [D19612337](https://our.internmc.facebook.com/intern/diff/D19612337/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit a95db14f:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=a95db14fb666a603c86e5ad787cead8fe48a78a2).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4365267) pytorch_libtorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92295576) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92295576)) :snowflake:
<details>
<summary>
<code>Jan 29 00:45:54 E: Failed to fetch https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1604/x86_64/./libnccl2_2.5.6-1+cuda10.1_amd64.deb  Operation too slow. Less than 10 bytes/sec transferred the last 120 seconds</code>
</summary>

```
Jan 29 00:41:38 ++ which ccache 
Jan 29 00:41:38 ++ '[' -z pytorch-libtorch-linux-xenial-cuda10.1-cudnn7-py3-gcc7-build ']' 
Jan 29 00:41:38 ++ [[ pytorch-libtorch-linux-xenial-cuda10.1-cudnn7-py3-gcc7-build == *pytorch-linux-xenial-cuda10.1-cudnn7-py3* ]] 
Jan 29 00:41:38 ++ [[ pytorch-libtorch-linux-xenial-cuda10.1-cudnn7-py3-gcc7-build == *pytorch-linux-trusty-py3.6-gcc7* ]] 
Jan 29 00:41:38 ++ [[ pytorch-libtorch-linux-xenial-cuda10.1-cudnn7-py3-gcc7-build == *pytorch_macos* ]] 
Jan 29 00:41:38 ++ BUILD_TEST_LIBTORCH=0 
Jan 29 00:41:38 ++ [[ pytorch-libtorch-linux-xenial-cuda10.1-cudnn7-py3-gcc7-build == *pytorch-linux-xenial-cuda* ]] 
Jan 29 00:41:38 + [[ pytorch-libtorch-linux-xenial-cuda10.1-cudnn7-py3-gcc7-build == *-xenial-cuda10.1-* ]] 
Jan 29 00:41:38 + sudo apt-get -qq update 
Jan 29 00:41:43 + sudo apt-get -qq install --allow-downgrades --allow-change-held-packages libnccl-dev=2.5.6-1+cuda10.1 libnccl2=2.5.6-1+cuda10.1 
Jan 29 00:45:54 E: Failed to fetch https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1604/x86_64/./libnccl2_2.5.6-1+cuda10.1_amd64.deb  Operation too slow. Less than 10 bytes/sec transferred the last 120 seconds 
Jan 29 00:45:54  
Jan 29 00:45:54 E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing? 
Jan 29 00:45:54 + cleanup 
Jan 29 00:45:54 + retcode=100 
Jan 29 00:45:54 + set +x 
Jan 29 00:45:54 =================== sccache compilation log =================== 
Jan 29 00:45:54 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 29 00:45:54 Compile requests                 0 
Jan 29 00:45:54 Compile requests executed        0 
Jan 29 00:45:54 Cache hits                       0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -60,25 +54,16 @@ py::object PyRRef::localValue() {
+  {
+    // acquiring GIL as torch::jit::toPyObject creates new py::object without
+    // grabbing the GIL.
+    pybind11::gil_scoped_acquire ag;
```
(can be done in followup PR) do we need to use `PROFILE_GIL_SCOPED_ACQUIRE` to profile this?

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -61,6 +60,7 @@ void PythonRpcHandler::cleanup() {
   pySerialize_ = py::none();
   pyHandleException_ = py::none();
   jitCompilationUnit_ = nullptr;
+  typeParser_ = nullptr;
```
Curious, it does not seem `ScriptTypeParser` depends on Python. Is setting it to nullptr here just being defensive or will it indeed hit error in certain destruction order?

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -181,15 +181,17 @@ PyObject* rpc_init(PyObject* /* unused */) {
+                    // __setstate__
+                    return PyRRef::unpickle(t);
+                  }),
+              py::call_guard<py::gil_scoped_release>())
```
would I be correct if I assume this GIL acquirement is moved to `rref_impl.cpp`?

zhaojuanmao replied:
yes, but it could wait for pythonrpchandler to be constructed in another thread, and that thread is trying to aquire GIL, if we do not release GIL in unpickle, it could result in deadlock

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -183,8 +183,14 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
-        SerializedPyObj result = PythonRpcHandler::getInstance().serialize(
-            jit::toPyObject(rref->getValue()));
+        auto value = rref->getValue();
+        py::object pyValue;
```
Is creating `py::object` out of GIL is OK?

xush6528 replied:
It's ok.
`inc_ref()` is called in `py::object`'s copy constructor with a reference passed in.
Here, ssing default constructor is like `std::make_shared(nullptr)`.

mrshenli reviewed (approved):
LGTM! Thanks!

I just have some minor questions inline. 

Regarding the test failure, looks like `test_quantized_rnn` is cleared on master. Could you please try rebase to the current master to trigger tests again?

xush6528 commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -36,20 +36,14 @@ py::object PyRRef::toHere() {
-    if (rref_->isPyObj()) {
-      // UserRRef<py::object>::toHere() calls python_rpc_handler which acquires
-      // GIL.
-      return jit::toPyObject(
```
2 facts that help me understand `jit::toPyObject()` (which returns a `py::object`) requires GIL.

1. Constructing and destructing an py::object C++ object changes the Python object's ref count.
Ref: https://pybind11.readthedocs.io/en/stable/reference.html#_CPPv4N6object6objectERK6object

2. Changing Python objet's ref count requires acquiring GIL.
Ref: https://docs.python.org/3/c-api/init.html#thread-state-and-the-global-interpreter-lock

Combing them together, anywhere a py:object C++ object is present, the GIL should be acquired.

mrshenli reviewed (changes requested):
Given @xush6528's comment, we need to extend GIL to cover all occurrences of `py::object`?

xush6528 commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -43,6 +43,8 @@ RRefForkData::RRefForkData(
 
 py::tuple RRefForkData::toPyTuple() const {
+  // add GIL as it is contructing a py::object
+  pybind11::gil_scoped_acquire ag;
```
Do we need GIL for `RRefForkData::fromPyTuple` as well? or it has already acquired GIL, how?

zhaojuanmao replied:
sounds good, let me add GIL for fromPyTuple as well

mrshenli replied:
it's a const reference of the `py::tuple` and does not seem to create/delete `py::object`. Is it still necessary to use GIL in `fromPyTuple`.

zhaojuanmao replied:
I guess GIL here can help avoid underneath python object pointed by pyTuple could be mutated?

xush6528 replied:
@mrshenli Accessing (both read and write) a Python object also requires GIL, otherwise, other thread could get GIL and invalidate that Python object.

zhaojuanmao commented:
@mrshenli I think we added GIL for all py::objects after adding GIL to fromPyTuple

zhaojuanmao committed:
Update on "clean up GIL usuage"


This is to follow up PR #30630, we need to have GIL when calling jit::toPyObject(), for some binded functions need to be taged with GIL release if underneath C++ codes requires GIL. so
1. pyRef::to_here() and pyRef::local_value() added GIL
2. pyRef::pickle and pyRef::unpickle() added GIL release tag
3. in request_callback_impl, also added GIL as needed
4. for typeParser, use cached jitCompilationUnit_, also clean it up in cleanUp() function

Differential Revision: [D19612337](https://our.internmc.facebook.com/intern/diff/D19612337/)

[ghstack-poisoned]

xush6528 reviewed (approved):
It's good to go now.
